lib/Common/DataStructures/WeakReferenceDictionary.h

```diff
@@ -44,4 +44,668 @@ namespace JsUtil
+        {
+            Assert(index >= 0);
+            Assert(index < count);
+
+            return values[index];
+        }
+
+        TKey const& GetKeyAt(const int index) const
+        {
+            Assert(index >= 0);
+            Assert(index < count);
+
+            return entries[index].Key();
+        }
+
+        bool TryGetValueAt(int index, TValue * value) const
+        {
+            if (index >= 0 && index < count)
+            {
+                *value = values[index];
+                return true;
+            }
+            return false;
+        }
+
+        bool Remove(const TKey& key)
+        {
+            int i, last;
+            uint targetBucket;
+            if (FindEntryWithKey(key, &i, &last, &targetBucket))
+            {
+                RemoveAt(i, last, targetBucket);
+                return true;
+            }
+            return false;
+        }
+
+        // Returns whether the dictionary was resized or not
+        bool EnsureCapacity()
+        {
+            if (freeCount == 0 && count == size)
+            {
+                Resize();
+                return true;
+            }
+
+            return false;
+        }
+
+        int GetNextIndex()
+        {
+            if (freeCount != 0)
+            {
+                Assert(freeCount > 0);
+                Assert(freeList >= 0);
+                Assert(freeList < count);
+                return freeList;
+            }
+
+            return count;
+        }
+
+        int GetLastIndex()
+        {
+            return count - 1;
+        }
+
+        void LockResize()
+        {
+            __super::LockResize();
+        }
+
+        void UnlockResize()
+        {
+            __super::UnlockResize();
+        }
+
+    private:
+
+        template <typename TLookup>
+        static hash_t GetHashCodeWithKey(const TLookup& key)
+        {
+            // set last bit to 1 to avoid false positive to make hash appears to be a valid recycler address.
+            // In the same line, 0 should be use to indicate a non-existing entry.
+            return TAGHASH(Comparer<TLookup>::GetHashCode(key));
+        }
+
+        static hash_t GetHashCode(const TKey& key)
+        {
+            return GetHashCodeWithKey<TKey>(key);
+        }
+
+        static uint GetBucket(hash_t hashCode, int bucketCount, int modFunctionIndex)
+        {
+            return SizePolicy::GetBucket(UNTAGHASH(hashCode), bucketCount, modFunctionIndex);
+        }
+
+        uint GetBucket(hash_t hashCode) const
+        {
+            return GetBucket(hashCode, this->bucketCount, modFunctionIndex);
+        }
+
+        static bool IsFreeEntry(const EntryType &entry)
+        {
+            // A free entry's next index will be (-2 - nextIndex), such that it is always <= -2, for fast entry iteration
+            // allowing for skipping over free entries. -1 is reserved for the end-of-chain marker for a used entry.
+            return entry.next <= -2;
```

Comment:
Add symbolic names for magic numbers?

Reply:
Do you think it's also worth updating the magic numbers in `DictionaryBase`?

Reply:
Would be nice cleanup but I don't feel too strongly